test_rmw_implementation/test/test_service.cpp

```diff
@@ -0,0 +1,325 @@
+  EXPECT_EQ((unsigned)2, service_request.uint32_value);
+  rmw_reset_error();
+
+  const char * implementation_identifier = srv->implementation_identifier;
+  srv->implementation_identifier = "not-an-rmw-implementation-identifier";
+  ret = rmw_take_request(srv, &header, &service_request, &taken);
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, ret) << rmw_get_error_string().str;
+  EXPECT_EQ(false, service_request.bool_value);  // Verify post conditions
+  EXPECT_EQ((unsigned)1, service_request.uint8_value);
+  EXPECT_EQ((unsigned)2, service_request.uint32_value);
+  EXPECT_EQ(false, taken);
+  rmw_reset_error();
+  srv->implementation_identifier = implementation_identifier;
+
+  test_msgs__srv__BasicTypes_Request__fini(&service_request);
+}
+
+TEST_F(CLASSNAME(TestService, RMW_IMPLEMENTATION), send_reponse_with_bad_arguments) {
+  constexpr char service_name[] = "/test";
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+  test_msgs__srv__BasicTypes_Response service_response;
+  ASSERT_TRUE(test_msgs__srv__BasicTypes_Response__init(&service_response));
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__srv__BasicTypes_Response__fini(&service_response);
+  });
+  service_response.bool_value = false;
+  service_response.uint8_value = 1;
+  service_response.uint32_value = 2;
+  test_msgs__srv__BasicTypes_Request request;
+  ASSERT_TRUE(test_msgs__srv__BasicTypes_Request__init(&request));
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__srv__BasicTypes_Request__fini(&request);
+  });
+  request.bool_value = false;
+  request.uint8_value = 1;
+  request.uint32_value = 2;
+  int64_t sequence_number;
+  rmw_service_info_t header;
+  rmw_service_t * srv =
+    rmw_create_service(node, ts, service_name, &rmw_qos_profile_default);
+  ASSERT_NE(nullptr, srv) << rmw_get_error_string().str;
+  rmw_client_t * client =
+    rmw_create_client(node, ts, service_name, &rmw_qos_profile_default);
+  ASSERT_NE(nullptr, client) << rmw_get_error_string().str;
+
+  rmw_ret_t ret = rmw_send_request(client, &request, &sequence_number);
+  ASSERT_EQ(RMW_RET_OK, ret);
+
+  size_t number_of_services = 1u;
+  rmw_wait_set_t * wait_set = rmw_create_wait_set(&context, number_of_services);
+  ASSERT_NE(nullptr, wait_set) << rmw_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rmw_ret_t ret = rmw_destroy_wait_set(wait_set);
+    EXPECT_EQ(ret, RMW_RET_OK) << rcutils_get_error_string().str;
+  });
+  void * array[1];
+  array[0] = srv->data;
+  rmw_services_t srv_array;
+  srv_array.service_count = 1u;
+  srv_array.services = array;
+  rmw_time_t timeout;
+  timeout.sec = 1;
+  timeout.nsec = rmw_intraprocess_discovery_delay.count() * 100;
+  ret = rmw_wait(nullptr, nullptr, &srv_array, nullptr, nullptr, wait_set, &timeout);
+  ASSERT_EQ(RMW_RET_OK, ret);
+  ASSERT_NE(nullptr, srv_array.services[0]);
+
+  ret = rmw_send_response(nullptr, &header.request_id, &service_response);
```

Comment:
@Lobotuerk hmm, shoudn't we take the request? what are we waiting for if not?

Reply:
My thought process was, go through the pipeline enough so that 
```
* \pre Given `request_header` must be the one previously taken along with the ROS service
 *   request to which we reply.
``` 
can be considered to be true.

Reply:
Hmm, I'd advise we actually take the request instead of assuming it'll match. It puts less constraints on the actual implementation.

Reply:
Updated on b3a29ba